New incremental compilation fails when deleting a contributed subcomponent

Version 2.5.0-beta03


To reproduce

1. Add this to `sample/library`:
    ```
    @ContributesSubcomponent(Any::class, AppScope::class)
    interface DescriptionSubcomponent {
      fun fatherProvider(): FatherProvider
      fun motherProvider(): MotherProvider
    
    
      @ContributesSubcomponent.Factory
      interface Factory {
        fun create(): DescriptionSubcomponent
      }
    
      @ContributesTo(AppScope::class)
      interface ParentComponent {
        fun descriptionSubcomponentFactory(): Factory
      }
    }
    ```
2. Run `./gradlew :delegate:sample:app:assembleDebug -Pcom.squareup.anvil.trackSourceFiles=true`
3. Remove the contributed subcomponent again
4. Run `./gradlew :delegate:sample:app:assembleDebug -Pcom.squareup.anvil.trackSourceFiles=true`

Step 4 fails with this error
```
> Task :delegate:sample:app:kaptDebugKotlin FAILED
/Users/gabrielittner/Downloads/anvil/sample/app/build/tmp/kapt3/stubs/debug/anvil/component/com/squareup/anvil/sample/appcomponent/DescriptionSubcomponentA.java:30: error: @Binds methods' parameter type must be assignable to the return type - anvil.component.com.squareup.anvil.sample.appcomponent.DescriptionSubcomponentA.SubcomponentModule.bindSubcomponentFactory(anvil.component.com.squareup.anvil.sample.appcomponent.DescriptionSubcomponentA.SubcomponentFactory)
        public abstract error.NonExistentClass bindSubcomponentFactory(@org.jetbrains.annotations.NotNull()
                                               ^
/Users/gabrielittner/Downloads/anvil/sample/app/build/tmp/kapt3/stubs/debug/anvil/component/com/squareup/anvil/sample/appcomponent/DescriptionSubcomponentA.java:17: error: @Subcomponent.Factory types must have exactly one method that returns the @Subcomponent type - anvil.component.com.squareup.anvil.sample.appcomponent.DescriptionSubcomponentA.SubcomponentFactory
    public static abstract interface SubcomponentFactory {
```

It looks like the code in the library module is properly cleaned up but the code that the app module generates isn't. Deleting `sample/app/build` and then running again fixes the issue.

--- a/compiler/src/main/java/com/squareup/anvil/compiler/Utils.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/Utils.kt
@@ -140,3 +140,35 @@ internal fun File.requireDelete() {
     throw AnvilCompilationException("Could not delete file: $this")
   }
 }
+
+/**
+ * Transforms the elements of the receiver collection and adds the results to a set.
+ *
+ * @param destination The destination set where the transformed
+ *   elements are placed. By default, it is an empty mutable set.
+ * @param transform Maps elements of the receiver collection to the output set.
+ * @receiver The collection to be transformed.
+ * @return A set containing the transformed elements from the receiver collection.
+ */
+internal inline fun <C : Collection<T>, T, R> C.mapToSet(
+  destination: MutableSet<R> = mutableSetOf(),
+  transform: (T) -> R,
+): Set<R> {
+  return mapTo(destination, transform)
+}
+
+/**
+ * Transforms the elements of the receiver array and adds the results to a set.
+ *
+ * @param destination The destination set where the transformed
+ *   elements are placed. By default, it is an empty mutable set.
+ * @param transform Maps elements of the receiver collection to the output set.
+ * @receiver The array to be transformed.
+ * @return A set containing the transformed elements from the receiver array.
+ */
+internal inline fun <T, R> Array<T>.mapToSet(
+  destination: MutableSet<R> = mutableSetOf(),
+  transform: (T) -> R,
+): Set<R> {
+  return mapTo(destination, transform)
+}

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/CodeGenerationExtension.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/CodeGenerationExtension.kt
@@ -11,6 +11,7 @@ import com.squareup.anvil.compiler.codegen.incremental.GeneratedFileCache
 import com.squareup.anvil.compiler.codegen.incremental.GeneratedFileCache.Companion.binaryFile
 import com.squareup.anvil.compiler.codegen.incremental.ProjectDir
 import com.squareup.anvil.compiler.codegen.reference.RealAnvilModuleDescriptor
+import com.squareup.anvil.compiler.mapToSet
 import com.squareup.anvil.compiler.requireDelete
 import org.jetbrains.kotlin.analyzer.AnalysisResult
 import org.jetbrains.kotlin.analyzer.AnalysisResult.RetryWithAdditionalRoots
@@ -82,7 +83,12 @@ internal class CodeGenerationExtension(
     val createdChanges = if (shouldRestoreFromCache()) {
       // For incremental compilation: restore any missing generated files from the cache,
       // and delete any generated files that are out-of-date due to source changes.
-      cacheOperations.restoreFromCache(generatedDir)
+      cacheOperations.restoreFromCache(
+        generatedDir = generatedDir,
+        inputKtFiles = files.mapToSet {
+          AbsoluteFile(File(it.virtualFilePath)).relativeTo(projectDir)
+        },
+      )
         .let { it.restoredFiles.isNotEmpty() || it.deletedFiles.isNotEmpty() }
     } else {
       // OLD incremental behavior: just delete everything if we're not tracking source files
@@ -118,6 +124,7 @@ internal class CodeGenerationExtension(
     bindingTrace: BindingTrace,
     files: Collection<KtFile>,
   ): AnalysisResult? {
+
     if (didRecompile) return null
     didRecompile = true
 

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/incremental/FileCacheOperations.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/incremental/FileCacheOperations.kt
@@ -2,6 +2,7 @@ package com.squareup.anvil.compiler.codegen.incremental
 
 import com.squareup.anvil.compiler.api.FileWithContent
 import com.squareup.anvil.compiler.api.GeneratedFileWithSources
+import com.squareup.anvil.compiler.mapToSet
 import com.squareup.anvil.compiler.requireDelete
 import java.io.File
 import java.io.Serializable
@@ -35,24 +36,40 @@ internal class FileCacheOperations(
 
   /**
    * For any non-generated file in the cache that no longer exists on disk,
-   * delete any downstream generated files from the disk.
+   * we delete any downstream generated files from the disk.
    *
-   * For any non-generated file on disk that is identical to its version in the cache,
-   * restore any downstream generated files from the cache to the disk.
+   * The [inputKtFiles] parameter should correspond to the `files: List<KtFile>` parameters in
+   * [doAnalysis][com.squareup.anvil.compiler.codegen.CodeGenerationExtension.doAnalysis].
+   *
+   * For any file in [inputKtFiles], we delete it and everything generated from it from the cache,
+   * and we delete anything generated from it from the disk.
+   * This is because we'll be passing those files through the code generators again anyway,
+   * so we'll get up-to-date results. This is also useful when a file's content is unchanged,
+   * but there was a classpath change that would result in different bindings or interface merges.
+   *
+   * For any non-generated file on disk that was *not* passed in as `inputKtFiles`
+   * *and* is identical to its version in the cache,
+   * we restore any downstream generated files from the cache to the disk.
    *
    * We use all cached source files as the basis for cache restoration,
-   * instead of using the `List<KtFile>` passed in by the compiler.
-   * The file list from the compiler can be incremental, not including all extant source files.
-   * With the incomplete list, any missing files
-   * that were generated from an unchanged source file would not be restored.
+   * instead of using the [inputKtFiles] passed in by the compiler.
+   * Those `inputKtFiles` are an incremental, possibly partial list of source files.
+   * If a source file exists on disk but isn't in the list,
+   * that means the compiler expects it and any generated files to be up-to-date --
+   * so that's what we need to restore.
    */
-  fun restoreFromCache(generatedDir: File): RestoreCacheResult {
+  fun restoreFromCache(generatedDir: File, inputKtFiles: Set<RelativeFile>): RestoreCacheResult {
 
     // Files that weren't generated.
     val rootSourceFiles = cache.rootSourceFiles
 
     val (changedSourceFiles, unchangedSourceFiles) = rootSourceFiles
-      .partition { cache.hasChanged(it) }
+      .partition {
+        // We treat anything in `inputKtFiles` as changed.
+        // If our cache is somehow out of sync with the file's content,
+        // then we can't trust anything else in the cache related to this file either.
+        it in inputKtFiles || cache.hasChanged(it)
+      }
 
     val deletedInvalid = deleteInvalidFiles(
       rootSourceFiles = rootSourceFiles,
@@ -102,7 +119,7 @@ internal class FileCacheOperations(
     generatedDir: File,
     validGeneratedFiles: Collection<FileWithContent>,
   ): List<AbsoluteFile> {
-    val validFilesAsFiles = validGeneratedFiles.mapTo(mutableSetOf()) { it.file }
+    val validFilesAsFiles = validGeneratedFiles.mapToSet { it.file }
 
     return generatedDir.walkBottomUp()
       .filter { it.isFile }

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/incremental/GeneratedFileCache.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/incremental/GeneratedFileCache.kt
@@ -4,6 +4,7 @@ import com.squareup.anvil.compiler.api.AnvilCompilationException
 import com.squareup.anvil.compiler.api.GeneratedFileWithSources
 import com.squareup.anvil.compiler.codegen.incremental.collections.Multimap
 import com.squareup.anvil.compiler.codegen.incremental.collections.MutableBiMap
+import com.squareup.anvil.compiler.mapToSet
 import java.io.Closeable
 import java.io.File
 import java.io.IOException
@@ -217,7 +218,7 @@ internal class GeneratedFileCache private constructor(
 
     fun getSourceFilesWithoutGeneratedContent(): Set<RelativeFile> = sourcesToGenerated.keys
       .filter { !generatedToContent.containsKey(it) }
-      .mapTo(mutableSetOf()) { fileIds.inverse.getValue(it) }
+      .mapToSet { fileIds.inverse.getValue(it) }
 
     fun putSourceGeneratedPair(sourceFile: RelativeFile, generatedFile: RelativeFile) {
       sourcesToGenerated.add(sourceFile.id, generatedFile.id)
@@ -248,11 +249,11 @@ internal class GeneratedFileCache private constructor(
 
     fun getSourceFiles(generatedFile: RelativeFile): Set<RelativeFile> =
       generatedToSources[generatedFile.id]
-        .mapTo(mutableSetOf()) { fileIds.inverse.getValue(it) }
+        .mapToSet { fileIds.inverse.getValue(it) }
 
     fun getGeneratedFiles(sourceFile: RelativeFile): Set<RelativeFile> =
       sourcesToGenerated[sourceFile.id]
-        .mapTo(mutableSetOf()) { fileIds.inverse.getValue(it) }
+        .mapToSet { fileIds.inverse.getValue(it) }
 
     fun putGeneratedContent(generatedFile: RelativeFile, content: String): String? {
       return generatedToContent.put(generatedFile.id, content)

--- a/compiler/src/test/java/com/squareup/anvil/compiler/codegen/incremental/CacheTestEnvironment.kt
+++ b/compiler/src/test/java/com/squareup/anvil/compiler/codegen/incremental/CacheTestEnvironment.kt
@@ -7,6 +7,7 @@ import com.rickbusarow.kase.NoParamTestEnvironmentFactory
 import com.rickbusarow.kase.files.TestLocation
 import com.rickbusarow.kase.stdlib.createSafely
 import com.squareup.anvil.compiler.api.GeneratedFileWithSources
+import com.squareup.anvil.compiler.mapToSet
 import io.kotest.matchers.file.shouldExist
 import io.kotest.matchers.shouldBe
 import java.io.File
@@ -120,6 +121,18 @@ internal class CacheTestEnvironment(
 
   fun FileCacheOperations.addToCache(vararg gens: GeneratedFileWithSources) {
     addToCache(emptyList(), gens.toList())
+  }
+
+  fun FileCacheOperations.restoreFromCache(vararg sources: FileType) {
+    restoreFromCache(
+      generatedDir = generatedDir,
+      inputKtFiles = sources.mapToSet { fileType ->
+        when (fileType) {
+          is AbsoluteFile -> fileType.relativeTo(projectDir)
+          is RelativeFile -> fileType
+        }
+      },
+    )
   }
 
   fun newCache(binaryFile: File = this.binaryFile): GeneratedFileCache {

--- a/compiler/src/test/java/com/squareup/anvil/compiler/codegen/incremental/FileCacheOperationsTest.kt
+++ b/compiler/src/test/java/com/squareup/anvil/compiler/codegen/incremental/FileCacheOperationsTest.kt
@@ -23,6 +23,30 @@ internal class FileCacheOperationsTest : CacheTests {
   }
 
   @Test
+  fun `restoreFromCache passing a source file as input deletes it and everything downstream`() = test {
+
+    fileOperations.addToCache(
+      gen1.withSources(source1),
+      gen2.withSources(source2),
+      gen3.withSources(gen2.toSourceFile()),
+      gen4.withSources(source3),
+    )
+
+    listOf(gen1, gen2, gen3)
+      .forEach { it.file.delete() }
+
+    fileOperations.restoreFromCache(generatedDir, setOf(source2))
+
+    currentFiles() shouldBe listOf(
+      gen1.relativeFile,
+      gen4.relativeFile,
+      source1,
+      source2,
+      source3,
+    )
+  }
+
+  @Test
   fun `restoreFromCache missing generated files are restored`() = test {
 
     fileOperations.addToCache(
@@ -34,7 +58,7 @@ internal class FileCacheOperationsTest : CacheTests {
     listOf(gen1, gen2, gen3)
       .forEach { it.file.delete() }
 
-    fileOperations.restoreFromCache(generatedDir)
+    fileOperations.restoreFromCache()
 
     currentFiles() shouldBe listOf(
       gen1.relativeFile,
@@ -64,7 +88,7 @@ internal class FileCacheOperationsTest : CacheTests {
       gen3.withSources(gen2.toSourceFile()),
     )
 
-    fileOperations.restoreFromCache(generatedDir)
+    fileOperations.restoreFromCache()
 
     currentFiles() shouldBe listOf(
       gen1.relativeFile,
@@ -87,7 +111,7 @@ internal class FileCacheOperationsTest : CacheTests {
 
       source1.writeText("changed")
 
-      fileOperations.restoreFromCache(generatedDir)
+      fileOperations.restoreFromCache(source1)
 
       currentFiles() shouldBe listOf(gen3.relativeFile, source1, source2)
     }
@@ -103,7 +127,7 @@ internal class FileCacheOperationsTest : CacheTests {
 
     source2.writeText("changed")
 
-    fileOperations.restoreFromCache(generatedDir)
+    fileOperations.restoreFromCache(source2)
 
     currentFiles() shouldBe listOf(gen1.relativeFile, source1, source2)
   }
@@ -121,7 +145,7 @@ internal class FileCacheOperationsTest : CacheTests {
 
     currentFiles() shouldBe listOf(gen1.relativeFile, untracked, source1)
 
-    fileOperations.restoreFromCache(generatedDir)
+    fileOperations.restoreFromCache()
 
     currentFiles() shouldBe listOf(gen1.relativeFile, source1)
   }
@@ -143,7 +167,7 @@ internal class FileCacheOperationsTest : CacheTests {
     )
       .forEach { it.absoluteFile.delete() }
 
-    fileOperations.restoreFromCache(generatedDir)
+    fileOperations.restoreFromCache()
 
     currentFiles() shouldBe listOf(gen2.relativeFile, gen3.relativeFile, source2)
   }

--- a/gradle-plugin/src/gradleTest/java/com/squareup/anvil/plugin/IncrementalTest.kt
+++ b/gradle-plugin/src/gradleTest/java/com/squareup/anvil/plugin/IncrementalTest.kt
@@ -1,6 +1,7 @@
 package com.squareup.anvil.plugin
 
 import com.rickbusarow.kase.Kase1
+import com.rickbusarow.kase.asClueCatching
 import com.rickbusarow.kase.files.DirectoryBuilder
 import com.rickbusarow.kase.gradle.dsl.buildFile
 import com.rickbusarow.kase.gradle.getValue
@@ -12,8 +13,10 @@ import com.rickbusarow.kase.wrap
 import com.squareup.anvil.plugin.testing.AnvilGradleTestEnvironment
 import com.squareup.anvil.plugin.testing.BaseGradleTest
 import com.squareup.anvil.plugin.testing.allBoundTypes
+import com.squareup.anvil.plugin.testing.allMergedModulesForComponent
 import com.squareup.anvil.plugin.testing.anvil
 import com.squareup.anvil.plugin.testing.classGraphResult
+import com.squareup.anvil.plugin.testing.names
 import io.kotest.matchers.file.shouldExist
 import io.kotest.matchers.file.shouldNotExist
 import io.kotest.matchers.shouldBe
@@ -555,6 +558,158 @@ class IncrementalTest : BaseGradleTest() {
       app.classGraphResult(lib).allBoundTypes() shouldBe listOf(
         "com.squareup.test.lib.OtherClass" to "com.squareup.test.lib.TypeB",
       )
+    }
+
+  @TestFactory
+  fun `a generated merged Subcomponent is deleted when a dependency module's @ContributesSubcomponent type is deleted`() =
+    testFactory {
+
+      // repro for https://github.com/square/anvil/issues/898
+
+      rootProject {
+        gradlePropertiesFile(
+          """
+          com.squareup.anvil.trackSourceFiles=true
+          com.squareup.anvil.addOptionalAnnotations=true
+          """.trimIndent(),
+        )
+
+        settingsFileAsFile.appendText(
+          """
+
+          include(":lib")
+          include(":app")
+          """.trimIndent(),
+        )
+
+        project("lib") {
+          buildFile {
+            plugins {
+              kotlin("jvm")
+              id("com.squareup.anvil")
+            }
+            anvil {
+              generateDaggerFactories.set(true)
+            }
+            dependencies {
+              implementation(libs.dagger2.annotations)
+            }
+          }
+          dir("src/main/java") {
+
+            injectClass(packageName = "com.squareup.test.lib")
+
+            kotlinFile(
+              "com/squareup/test/lib/LibComponent.kt",
+              """
+              package com.squareup.test.lib
+
+              import com.squareup.anvil.annotations.ContributesTo
+              import com.squareup.anvil.annotations.optional.SingleIn
+
+              @ContributesTo(Any::class)
+              interface LibComponent {
+                fun injectClass(): InjectClass
+              }
+              """.trimIndent(),
+            )
+          }
+        }
+
+        project("app") {
+
+          buildFile {
+            plugins {
+              kotlin("jvm")
+              kotlin("kapt")
+              id("com.squareup.anvil")
+            }
+            dependencies {
+              compileOnly(libs.dagger2.annotations)
+              implementation(project(":lib"))
+              kapt(libs.dagger2.compiler)
+            }
+          }
+
+          dir("src/main/java") {
+            kotlinFile(
+              "com/squareup/test/app/AppComponent.kt",
+              """
+              package com.squareup.test.app
+
+              import com.squareup.anvil.annotations.MergeComponent
+              import com.squareup.anvil.annotations.optional.SingleIn
+
+              @SingleIn(Any::class)
+              @MergeComponent(Any::class)
+              interface AppComponent
+              """.trimIndent(),
+            )
+          }
+        }
+      }
+
+      val lib by rootProject.subprojects
+      val app by rootProject.subprojects
+
+      val subcomponentSrc = lib.path.resolve("src/main/java")
+        .resolve("com/squareup/test/lib/LibSubcomponent.kt")
+
+      "the first compilation happens with LibSubcomponent in the dependency module".asClueCatching {
+
+        subcomponentSrc.kotlin(
+          """
+          package com.squareup.test.lib
+
+          import com.squareup.anvil.annotations.ContributesTo
+          import com.squareup.anvil.annotations.ContributesSubcomponent
+
+          @ContributesSubcomponent(Unit::class, Any::class)
+          interface LibSubcomponent {
+            fun injectClass(): InjectClass
+            
+            @ContributesSubcomponent.Factory
+            interface Factory {
+              fun create(): LibSubcomponent
+            }
+
+            @ContributesTo(Any::class)
+            interface ParentComponent {
+              fun libSubcomponentFactory(): Factory
+            }
+          }
+          """.trimIndent(),
+        )
+
+        shouldSucceed("jar")
+
+        app.classGraphResult(lib).apply {
+
+          allMergedModulesForComponent("com.squareup.test.app.AppComponent")
+            .names() shouldBe listOf(
+            "anvil.module.com.squareup.test.app.AppComponentAnvilModule",
+            "anvil.component.com.squareup.test.app.appcomponent.LibSubcomponentA\$SubcomponentModule",
+          )
+        }
+      }
+
+      "The second compilation happens after deleting LibSubcomponent".asClueCatching {
+
+        subcomponentSrc.deleteOrFail()
+
+        shouldSucceed("jar") {
+          task(":lib:compileKotlin")?.outcome shouldBe TaskOutcome.SUCCESS
+          task(":app:compileKotlin")?.outcome shouldBe TaskOutcome.SUCCESS
+        }
+
+        app.classGraphResult(lib).apply {
+
+          allMergedModulesForComponent("com.squareup.test.app.AppComponent")
+            .names() shouldBe listOf(
+            "anvil.module.com.squareup.test.app.AppComponentAnvilModule",
+          )
+        }
+      }
     }
 
   @TestFactory

--- a/gradle-plugin/src/gradleTest/java/com/squareup/anvil/plugin/testing/AnvilGradleTestEnvironment.kt
+++ b/gradle-plugin/src/gradleTest/java/com/squareup/anvil/plugin/testing/AnvilGradleTestEnvironment.kt
@@ -51,9 +51,9 @@ class AnvilGradleTestEnvironment(
    *   )
    * ```
    */
-  fun File.listRelativeFilePaths(): List<String> = walkBottomUp()
+  fun File.listRelativeFilePaths(parentDir: File = this): List<String> = walkBottomUp()
     .filter { it.isFile }
-    .map { it.toRelativeString(rootAnvilMainGenerated) }
+    .map { it.toRelativeString(parentDir) }
     .sorted()
     .toList()
 

